src/new_index/mempool.rs

```diff
@@ -577,6 +503,68 @@ impl Mempool {
             .get(asset_id)
             .map_or_else(|| vec![], |entries| self._history(entries, limit))
     }
+
+    pub fn download_new_mempool_txs(
+        daemon: &Daemon,
+        old_txids: &HashSet<Txid>,
+        new_txids: &HashSet<Txid>
+    ) -> Result<Vec<Transaction>>{
+        let t = Instant::now();
+
+        let txids: Vec<&Txid> = (*new_txids).difference(old_txids).collect();
+        let transactions = match daemon.gettransactions(&txids) {
+            Ok(txs) => txs,
+            Err(err) => {
+                warn!("failed to get {} transactions: {}", txids.len(), err); // e.g. new block or RBF
+                return Err(err);
+            }
+        };
+
+        log_fn_duration("mempool::download_new_mempool_txs", t.elapsed().as_micros());
+        return Ok(transactions);
+    }
+
+    pub fn update(mempool: &Arc<RwLock<Mempool>>, daemon: &Daemon) -> Result<()> {
+        let t = Instant::now();
+
+        // 1. Determine which transactions are no longer in the daemon's mempool
+        let old_txids = mempool.read().unwrap().old_txids();
+        let new_txids = daemon
+            .getmempooltxids()
+            .chain_err(|| "failed to update mempool from daemon")?;
+        let old_mempool_txs: HashSet<&Txid> = old_txids.difference(&new_txids).collect();
+        log_fn_duration("mempool::paratial_tx_fetch", t.elapsed().as_micros());
+
+        // 2. Download new transactions from the daemon's mempool
+        let new_mempool_txs = match Mempool::download_new_mempool_txs(&daemon, &old_txids, &new_txids) {
+            Ok(txs) => txs,
+            Err(_) => {
+                warn!("Failed to get new mempool txs, skipping mempool update");
+                return Ok(());
+            }
+        };
+
+        // 3. Update local mempool to match daemon's state
+        {
```

Comment:
The block scope isn't strictly necessary here since this is the last thing to run in the function. But I kind of like it that way anyway